volatility3/framework/plugins/windows/direct_system_calls.py

```diff
@@ -0,0 +1,455 @@
+                # then we are in an invalid path, so bail
+                if not syscall_finder.wants_syscall_inst or (
+                    not (found_movr10 and found_movreax)
+                ):
+                    break
+
+            else:
+                # attempt to see if any other instruction type wrote to registers
+                try:
+                    _, regs_written = inst.regs_access()
+                except capstone.CsError:
+                    continue
+
+                if regs_written:
+                    for r in regs_written:
+                        # track writes to eax/rax or R10
+                        reg = inst.reg_name(r)
+                        if reg in ["eax", "rax"]:
+                            found_movreax = True
+
+                        elif reg == "r10":
+                            found_movr10 = True
+
+        # if any of these are missing, the block is invalid regardless of
+        # the technique we are trying to detect now or in the future
+        if not (found_movr10 and found_movreax and found_end):
+            return None
+
+        # if the finder requires a 'syscall' instruction then bail now if we didn't find one
+        if syscall_finder.wants_syscall_inst and not found_syscall:
+            return None
+
+        return disasm_bytes, end_inst
+
+    @staticmethod
+    def get_disasm_function(architecture: str) -> Callable:
+        """
+        Returns the disassembly handler for the given architecture
+        .detail is used to get full instruction information
+
+        Args:
+            architecture: the name of the architecture for the process being disassembled
+        Returns:
+            The disasm function from capstone for the given architecture
+        """
+        disasm_types = {
+            "intel": capstone.Cs(capstone.CS_ARCH_X86, capstone.CS_MODE_32),
+            "intel64": capstone.Cs(capstone.CS_ARCH_X86, capstone.CS_MODE_64),
+        }
+
+        disasm_type = disasm_types[architecture]
+        disasm_type.detail = True
+        return disasm_type.disasm
+
+    @classmethod
+    def _is_valid_syscall(
+        cls,
+        syscall_finder: syscall_finder_type,
+        proc_layer: interfaces.layers.DataLayerInterface,
+        architecture: str,
+        vads: List[Tuple[int, int, str]],
+        address: int,
+    ) -> Optional[Tuple[int, str]]:
+        """
+        Args:
+            syscall_finder:
+            proc_layer: the memory layer of the process being scanned
+            architecture: the name of the architecture for the process being disassembled
+            vads: the ranges of this process under 10MB
+            address: the starting address to check for malicious syscall code blocks
+
+        Returns:
+            Optional[Tuple[int, str]]: For valid code blocks, the starting address of the block and the disassembly string
+        """
+        # the number bytes behind the yara rule hit to scan
+        behind = 32
```

Comment:
Can this not be encoded in the yara rule, or are we happy hard coding it here?  I guess it's about the same either way, just floating specific numbers get the hairs on the back of my neck up...  5;P

Reply:
Yes I much prefer it this way.